src/io/flutter/sdk/FlutterSdkUtil.java

```diff
@@ -213,4 +219,73 @@ public static void enableDartSdk(@NotNull final Project project) {
       DartPlugin.enableDartSdk(modules[0]);
     }
   }
+
+  /**
+   * Parse any .packages file and infer the location of the Flutter SDK from that.
+   */
+  @Nullable
+  public static String guessFlutterSdkFromPackagesFile(@NotNull Module module) {
+    final PubRoot pubRoot = PubRoot.forModuleWithRefresh(module);
+    if (pubRoot == null) {
+      return null;
+    }
+
+    final VirtualFile packagesFile = pubRoot.getPackages();
+    if (packagesFile == null) {
+      return null;
+    }
+
+    // parse it
+    try {
+      final String contents = new String(packagesFile.contentsToByteArray(true /* cache contents */));
+      return parseFlutterSdkPath(contents);
+    }
+    catch (IOException e) {
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  public static String parseFlutterSdkPath(String packagesFileContent) {
+    for (String line : packagesFileContent.split("\n")) {
+      // flutter:file:///Users/.../flutter/packages/flutter/lib/
+      line = line.trim();
+
+      if (line.isEmpty() || line.startsWith("#")) {
+        continue;
+      }
+
+      final String flutterPrefix = "flutter:";
+      if (line.startsWith(flutterPrefix)) {
+        final String urlString = line.substring(flutterPrefix.length());
+
+        if (urlString.startsWith("file:")) {
+          final Url url = Urls.parseEncoded(urlString);
+          if (url == null) {
+            continue;
+          }
+          final String path = url.getPath();
+          // go up three levels
+          final File file = new File(url.getPath());
+          return file.getParentFile().getParentFile().getParentFile().getPath();
+        }
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Locate the Flutter SDK using the user's PATH.
+   */
+  @Nullable
+  public static String locateSdkFromPath() {
+    final String flutterBinPath = Which.which("flutter");
```

Comment:
Which.which?
I'd have picked a different class name, maybe Sys, so other functions could be added.